R/webApiDataSources.R

```diff
@@ -0,0 +1,118 @@
+#'
+#' @examples
+#' \dontrun{
+#' # This will obtain pre-computed summary regarding dataSource
+#'
+#' getDataSourceReportForSourceKey(baseUrl = "http://server.org:80/WebAPI", dataSource = 'dataSource', reportType = 'dashboard')
+#' }
+#'
+#' @export
+
+getDataSourceReportForSourceKey <- function(baseUrl,
+                                            sourceKey,
+                                            reportType){
+  url <- sprintf("%s/cdmresults/%s/%s", baseUrl, sourceKey, reportType)
+  .getApiResponseParse(url)
+}
+
+#' Get the all the results from all Achilles reports for all provided data Sources
+#'
+#' @details
+#' Obtains the JSON expression from WebAPI for a list of dataSource, derived from given list of CDMschemas and returns list of tibble dataframes
+#'
+#' @param baseUrl        The base URL for the WebApi instance, for example:
+#'                       "http://server.org:80/WebAPI".
+#' @param sourceKeys    A list of source keys for a CDM instance in WebAPI, as defined in the
+#'                       Configuration page.
+#' @return
+#' A list of object representing pre computed Achilles results for the list of sourceKey's
+#'
+#' @examples
+#' \dontrun{
+#' # This will obtain pre-computed summary regarding dataSource
+#'
+#' getDataSource(baseUrl = "http://server.org:80/WebAPI", cdmDatabaseSchemas = c('schema1','schema2')
+#' }
+#'
+#' @export
+
+getDataSource <- function(baseUrl,sourceKeys){
+  require(dplyr)
+  .checkWebApiBaseUrl(baseUrl)
+
+  #validate source keys
+  sourceKeys <- .getValidSourceKeys(baseUrl, sourceKeys)
+
+  reportTypes <- c('dashboard', 'datadensity', 'person', 'visit', 'condition', 'conditionera',
+                      'procedure', 'drug', 'drugera', 'measurement', 'observation', 'death', 'achillesheel'
+                      )
+
+  writeLines(paste0('Downloading Achilles data into R'))
+  result <- list() #i = 1
+  for (i in (1:length(sourceKeys))) {
+    sourceKey = sourceKeys[[i]]
+    configuration <- getConfiguration(baseUrl = baseUrl,
+                                                       sourceKeys = sourceKey
+                                                    )$parsed
+
+    writeLines(paste0('Working on SourceKey = ', sourceKey))
+    for (j in (1:length(reportTypes))) {
+      reportType <- reportTypes[j]
+      writeLines(paste0('    Downloading = ', reportType))
+
+      result[[sourceKey]][[reportType]] <-
+        ROhdsiWebApi::getDataSourceReportForSourceKey(baseUrl = baseUrl
+                                                         , sourceKey = sourceKey
+                                                         , reportType = reportType
+                                                        )
+
+
+      jsonParsed <- result[[sourceKey]][[reportType]]$parsed
+
+      jsonNames <- names(jsonParsed)
+
+      if (is.null(jsonNames)) {
+        result[[sourceKey]][[reportType]]$parsed <- dplyr::bind_rows(jsonParsed)
+      } else {
+          for (k in (1:length(jsonNames))) {
```

Comment:
Triple-nested for-loop: consider vectorizing

Reply:
To Do (will attempt in future?)

Reply:
@schuemie do you mean using lapply?